### DISCUSSION TITLE
This QueryDict instance is immutable

### DISCUSSION THREAD

**Discussion by carthique** [netbox-community/netbox#13217] [Help Wanted!] [answered]
### NetBox version

v3.5.6

### Python version

3.10

### Steps to Reproduce

Trying to get available-ips using the below curl call
```
curl -X 'POST' \
  'http://localhost:8000/api/ipam/prefixes/5/available-ips/' \
  -H 'accept: application/json' \
  -H 'Content-Type: multipart/form-data' \
  -H 'X-CSRFTOKEN: LLaGfBUEU5uVlM4ghWXXC34HFsGp3rgF75UVsnXcx4jMlQ5ADLC9o8tSkNqOHS67' \
  -F 'address=10.0.0.0/16'
```

We get the following error

```
<pre class="block"><strong>&lt;class &#x27;AttributeError&#x27;&gt;</strong><br />
This QueryDict instance is immutable
```

### Expected Behavior

To get the list of IP addresses available
`[10.0.0.2/16, 10.0.0.4/16, 10.0.0.5/16, 10.0.0.6/16, 10.0.0.7/16, 10.0.0.8/16, 10.0.0.9/16, 10.0.0.10/16, 10.0.0.11/16, 10.0.0.12/16, 10.0.0.13/16, 10.0.0.14/16, 10.0.0.15/16, 10.0.0.16/16, 10.0.0.17/16, 10.0.0.18/16, 10.0.0.19/16, 10.0.0.20/16, 10.0.0.21/16, 10.0.0.22/16, 10.0.0.23/16, 10.0.0.24/16, 10.0.0.25/16, 10.0.0.26/16, 10.0.0.27/16, 10.0.0.28/16, 10.0.0.29/16, 10.0.0.30/16, 10.0.0.31/16, 10.0.0.32/16, 10.0.0.33/16, 10.0.0.34/16, 10.0.0.35/16, 10.0.0.36/16, 10.0.0.37/16, 10.0.0.38/16, 10.0.0.39/16, 10.0.0.40/16, 10.0.0.41/16, 10.0.0.42/16, 10.0.0.43/16, 10.0.0.44/16, 10.0.0.45/16, 10.0.0.46/16, 10.0.0.47/16, 10.0.0.48/16, 10.0.0.49/16, 10.0.0.50/16, 10.0.0.51/16, 10.0.0.52/16]
`

### Observed Behavior

```
<!DOCTYPE html>
<html lang="en">

<head>
    <title>Server Error</title>
    <link rel="stylesheet" href="/static/netbox-light.css" />
    <meta charset="UTF-8">
</head>

<body>
    <div class="container-fluid">
        <div class="row">
            <div class="col col-md-6 offset-md-3">
                <div class="card border-danger mt-5">
                    <h5 class="card-header">
                        <i class="mdi mdi-alert"></i> Server Error
                    </h5>
                    <div class="card-body">
                        
                            <p>
                                There was a problem with your request. Please contact an administrator.
                            </p>
                        
                        <hr />
                        <p>
                            The complete exception is provided below:
                        </p>
<pre class="block"><strong>&lt;class &#x27;AttributeError&#x27;&gt;</strong><br />
This QueryDict instance is immutable

Python version: 3.10.6
NetBox version: 3.5.6</pre>
                        <p>
                            If further assistance is required, please post to the <a href="https://github.com/netbox-community/netbox/discussions">NetBox discussion forum</a> on GitHub.
                        </p>
                        <div class="text-end">
                            <a href="/" class="btn btn-primary">Home Page</a>
                        </div>
                    </div>
                </div>
            </div>
        </div>
    </div>
</body>

</html>
```

**Comment by abhi1693** [comment, marked as answer]
Please review your API call as it does not comply with the requirements: https://demo.netbox.dev/api/schema/swagger-ui/#/ipam/ipam_prefixes_available_ips_create

There are a few issues here:

1. You have not provided the Autorization token
2. The endpoint does not require any payload
3. The endpoint requires json content type

**Comment by candlerb** [comment]
If you want to "get the list of IP addresses available" then you need to make a GET request, not a POST request - and it works fine for me.

There is no need to provide the body parameter `address=10.0.0.0/16`. The address of the prefix is implicit from the prefix ID (that is, `/api/ipam/prefixes/5/available-ips/` is referring to the prefix with id=5)

**Reply by carthique** [reply]
Sorry @candlerb, my bad with post what I am trying to do here is to get available ip from the pool. Expectation is to get a /32 ip address from the pool.

**Reply by candlerb** [reply]
As I say, it works for me.

```
$ export NETBOX=https://....
$ export TOKEN=...
$ curl -sS -H "Authorization: Token $TOKEN" "$NETBOX/api/ipam/prefixes/1/available-ips/" | jq .
[
  {
    "family": 4,
    "address": "10.12.254.6/24",
    "vrf": null
  },
  {
    "family": 4,
    "address": "10.12.254.7/24",
    "vrf": null
  },
  {
    "family": 4,
    "address": "10.12.254.8/24",
    "vrf": null
  },
  {
    "family": 4,
    "address": "10.12.254.9/24",
    "vrf": null
  },
... etc
```

**Reply by carthique** [reply]
Awesome! I think I forgot to refresh the auth token and that is the issue. But how do I create/allocate a /32 ip address from the pool of available ips within the prefix.

**Reply by candlerb** [reply]
At simplest, you could just pick one of these IPs and create an IP address object with a regular POST.

It may be possible to assign one atomically using a POST to available-ips, but I can't see the semantics documented anywhere. The [auto-generated swagger API doc](https://demo.netbox.dev/api/schema/swagger-ui/#/ipam/ipam_prefixes_available_ips_create) is not helpful: it just says "Post a IP address object." and shows a random-looking request body.

I'm afraid this a common problem with REST APIs: the format of a request may be given, but the semantics are not.

**Reply by carthique** [reply]
Yeah, for picking the ip, i need it to be atomic op as there could be multiple requests coming in at the same time. Let me try out the post call with auth header and see if it returns a /32. I am able to get this done using the pynetbox library. Want the API version of the same. I really appreciate all the help I am getting here. Thanks

**Reply by candlerb** [reply]
> I am able to get this done using the pynetbox library. Want the API version of the same

Then just use tcpdump to show what pynetbox is sending. Do this between Apache/Nginx and Gunicorn, so you can see the plaintext traffic.

```
tcpdump -i lo -nn -s0 -A tcp port 8001
```

**Reply by carthique** [reply]
I tried with authentication and it works. Although it only works with content type application/json. I still hit the query dist is immutable error when using multipart/form-data may be an issue with swagger API doc.
Anyways, this solves my issue, thanks  a lot for the help @candlerb and @abhi1693

**Comment by mtinberg** [comment]
I just want to be clear, that IP Address objects can have the real netmask associated and don't need to be created with /32, you can tell they are a single address because they are an ip-address and not a prefix, to make templating of resulting configs easier as you don't need to look up the containing prefix to get the mask (might still need it to get the gateway if it isn't something you can assume by convention in your templates).

So you'd create a record for { "address": "10.12.254.6/24" } and not 10.12.254.6/32 unless you really wanted to strip this context from ip-address records (eg. I don't worry about netmask for VIPs/Anycast ip-address records)


—

Mark Tinberg ***@***.***>

Division of Information Technology-Network Services

University of Wisconsin-Madison


________________________________

**Reply by carthique** [reply]
Thanks @mtinberg but I am trying to get a host(/32) IP address from the prefix here so I need a /32 and not a subnet.

**Reply by candlerb** [reply]
When Netbox says "10.12.254.6/24" it's actually a single IP.  That is, it means: "the single IP address 10.12.254.6 (/32 implied), but for use on a network with a prefix length of /24 (which is 10.12.254.0/24, also implied)"

That's just how it is. Money back if not satisfied.

**Reply by carthique** [reply]
@candlerb just realized that when I saw the response, you can keep the money 😄 as I am fully satisfied. Thanks for all the effort in helping me out with this

`{"id":641,"url":"http://localhost:8000/api/ipam/ip-addresses/641/","display":"10.0.0.11/16","family":{"value":4,"label":"IPv4"},"address":"10.0.0.11/16","vrf":null,"tenant":null,"status":{"value":"active","label":"Active"},"role":null,"assigned_object_type":null,"assigned_object_id":null,"assigned_object":null,"nat_inside":null,"nat_outside":[],"dns_name":"","description":"","comments":"","tags":[],"custom_fields":{"allow_listed":true,"city":null,"country":null,"ip_type":null,"location":null},"created":"2023-07-19T19:20:23.942613Z","last_updated":"2023-07-19T19:20:23.942634Z"}`
